rskj-core/src/main/java/co/rsk/trie/Trie.java

```diff
@@ -766,23 +766,18 @@ private Trie put(TrieKeySlice key, byte[] value, boolean isRecursiveDelete) {
             return trie;
         }
 
-        Optional<Trie> leftOpt = trie.left.getNode();
-        Optional<Trie> rightOpt = trie.right.getNode();
-        if (leftOpt.isPresent() && rightOpt.isPresent()) {
-            return trie;
-        }
-
-        if (!leftOpt.isPresent() && !rightOpt.isPresent()) {
+        // both left and right exist (or not) at the same time
+        if (trie.left.isEmpty() == trie.right.isEmpty()) {
             return trie;
         }
 
         Trie child;
         byte childImplicitByte;
-        if (leftOpt.isPresent()) {
-            child = leftOpt.get();
+        if (!trie.left.isEmpty() && trie.left.getNode().isPresent()) {
```

Comment:
My `null` guard comment was regarding `trie.left`, not `trie.left.getNode()`, which will never return `null` as it always returns an `Optional`. This guard calls `getNode()` which is quite expensive unnecessarily (as it has to perform `getHash()`)

This should be taken with a grain of salt though, even though it is possible that `trie.left` is null (nothing in the code prevents it as far as I see), the old code actually does assume that it is not null, maybe it is best to keep it that way.

So:

* To avoid risks of changing behavior I'd keep `trie.left`, `trie.right` without null checks, as the previous code assumed they were always something and this is most likely what's happening if there's no reason to believe otherwise
* The added guard adds more computation and also very subtly modifies behavior.

Reply:
Previously the code was doing the checkup about the node being not null here:

```java
Optional<Trie> leftOpt = trie.left.getNode();
Optional<Trie> rightOpt = trie.right.getNode();
if (leftOpt.isPresent() && rightOpt.isPresent()) { // <- Checking Here
    return trie;
}
```
so, in the following lines, when we called 

```java
rightOpt.get()
```

We were already covered.

But in the new code we only check that left and right are not empty without checking the contained node _per se_:

```java
if (trie.left.isEmpty() == trie.right.isEmpty()) {
```

So when calling

```java
.getNode().get()
```

we are not sure that the contents are not null.

So I think that:
- one of the points "To avoid risks of changing behavior I'd keep trie.left, trie.right without null checks, as the previous code assumed they were always something..." it's not completely right (because the old code was performing checks on the node being not null).
- we have to decide where and how we want to avoid extra computation. In this case the computation would be made for the cases when we know for sure that one of the left or right references are not empty.

What do you think?

Reply:
And yes `trie.left` and `trie.right` were not being checked for null right before as you pointed.

Reply:
With the checks removed, now if the `Node` element is not present, a `java.util.NoSuchElementException` is thrown by the `get()` method. So I've removed the checks for clarity, and we can start improving the solution from [this new commit](https://github.com/rsksmart/rskj/pull/1575/commits/2580b15ff41217a6abbc50d3bc643f043daae7e3) and on.

Reply:
Leaving for further discussion: the underlying issue here is that `isEmpty` and `getNode` are not correlated to what one would expect: an `!trie.isEmpty` **does not** assure that `trie.getNode().isPresent()`, at least according to the code.

IMHO this scenario should not be considered an `option`: looking up a node in the db which does not exist should be considered an exception, no correct alternative exists to continue processing the information.